doc/memory.rst

```diff
@@ -385,6 +385,23 @@ Gotchas
      ``self.method`` does not depend on ``self`` you can use
      ``self.method = memory.cache(self.method, ignore=['self'])``.
 
+* Here are important gotchas about ``joblib`` cache persistence across
+  environment upgrades:
+
+  1. **``joblib.Memory`` caches will get wiped out upon third-party library
+       updates**. If one updates any third-party library of an environment in
+       which ``joblib`` was installed and was maintaining caches, such caches
+       are likely to get invalidated.
+
+  2. **``joblib.Memory`` caches will persist across minor joblib version
+     updates**. If one updates ``joblib`` in an environment in which
+     ``joblib`` was installed and was maintaining caches, existing caches
+     will be conserved and used, unless the ``joblib`` **major** version
+     number changed.
+     Example: A ``joblib`` cache will survive a ``0.15.0`` to ``0.16.0``
+     ``joblib`` upgrade, but will not survive a ``0.15.0`` to ``1.0.0``
+     ``joblib`` upgrade.
```

Comment:
Since we have not easy to maintain way to ensure this contract (w.r.t. versioning semantics) I would rather just loosen the contract such that:


The values returned by `joblib.hash` are not guaranteed to stay constant when upgrading joblib to a new version and therefore the entries of a `joblib.Memory` cache can also get invalidated when upgrading as a result. Invalidation can also happen when the parameters of a cached function are instances of a classes defined in upgraded third party libraries such as numpy for instance.

Reply:
It is true that as of now, the tests I have written would not ensure such a contract is respected. I'm still a bit un-easy about the fact that upgrading `joblib` to for example pull a bugfix (something that may be necesary), can make user risk cache invalidation.

Maybe a intermediate would be to state best-effort hash stability across `joblib` releases.

But I'll leave it to you to have the last word on this.

Reply:
> Maybe a intermediate would be to state best-effort hash stability across joblib releases. 

I would rather say that most joblib upgrades are unlikely to result in a change of the hashing results but this is not guaranteed and can happen from time to time in particular when fixing bugs.